kmip/pie/client.py

```diff
@@ -0,0 +1,340 @@
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import logging
+import six
+
+from kmip.core.enums import AttributeType
+from kmip.core.enums import CryptographicAlgorithm
+from kmip.core.enums import CryptographicUsageMask
+from kmip.core.enums import ObjectType
+from kmip.core.enums import ResultStatus
+
+from kmip.core.factories.attributes import AttributeFactory
+
+from kmip.core.objects import CommonTemplateAttribute
+from kmip.core.objects import TemplateAttribute
+
+from kmip.pie.factory import ObjectFactory
+
+from kmip.pie.exceptions import ClientConnectionFailure
+from kmip.pie.exceptions import ClientConnectionNotOpen
+from kmip.pie.exceptions import KmipOperationFailure
+
+from kmip.pie.objects import Key
+
+from kmip.services.kmip_client import KMIPProxy
+
+
+class KmipClient:
+    """
+    A simplified PyKMIP client for conducting KMIP operations.
+
+    The KmipClient provides a simpler external interface for various KMIP
+    operations and composes the bulk of the PyKMIP Pie API. It wraps the
+    original KMIPProxy, reducing the boilerplate needed to deploy PyKMIP
+    in client applications.
+    """
+    def __init__(self,
+                 hostname=None,
+                 port=None,
+                 cert=None,
+                 key=None,
+                 ca=None,
+                 config='client'):
+        """
+        Construct a KMIPClient.
+
+        Args:
+            hostname (string): The host or IP address of a KMIP appliance.
+            port (int): The port number used to establish a connection to a
+                KMIP appliance. Usually 5696 for KMIP applications.
+            cert (string): The path to the client's certificate.
+            key (string): The path to the key for the client's certificate.
+            ca (string): The path to the CA certificate used to verify the
+                server's certificate.
+            config (string): The name of a section in the PyKMIP configuration
+                file. Use to load a specific set of configuration settings from
+                the configuration file, instead of specifying them manually.
+        """
+        self.logger = logging.getLogger()
+
+        self.attribute_factory = AttributeFactory()
+        self.object_factory = ObjectFactory()
+
+        self.proxy = KMIPProxy(
+            host=hostname,
+            port=port,
+            certfile=cert,
+            keyfile=key,
+            ca_certs=ca,
+            config=config)
+
+        self._is_open = False
+
+    def open(self):
+        """
+        Open the client connection.
+
+        Raises:
+            ClientConnectionFailure: if the client connection is already open
+            Exception: if an error occurs while trying to open the connection
+        """
+        if self._is_open:
```

Comment:
This does not seem thread safe. We should add a note at class level to indicate this. In the future if we want to add it then we can modify, but I do not have a requirement for this.